src/rgw/rgw_acl_s3.cc

```diff
@@ -366,8 +366,13 @@ int RGWAccessControlList_S3::create_canned(ACLOwner& owner, ACLOwner& bucket_own
   rgw_user bid = bucket_owner.get_id();
   string bname = bucket_owner.get_display_name();
 
-  /* owner gets full control */
-  owner_grant.set_canon(owner.get_id(), owner.get_display_name(), RGW_PERM_FULL_CONTROL);
+  /* owner gets full control, while anonymous putting object in
+   * public-write bucket should belong it to the bucket owner*/
+  if(owner.get_id() == rgw_user(RGW_USER_ANON_ID)) {
+    owner_grant.set_canon(bid, bname, RGW_PERM_FULL_CONTROL);
+  } else {
+    owner_grant.set_canon(owner.get_id(), owner.get_display_name(), RGW_PERM_FULL_CONTROL);
+  }
```

Comment:
this is better, yeah. but for the 'bucket-owner-read' and 'bucket-owner-full-control' cases below, we're still making comparisons against the original `owner`, so we would end up with two separate grants for the bucket owner

maybe it is better for `create_canned()` to take `ACLOwner owner` by value and overwrite it, so those 'bucket-owner-' comparisons don't have to change

Reply:
Sorry for less test for last modification. Today when I test for the acl owner, I found out that if we do not change the `owner` with reference int 'RGWAccessControlList_S3::create_canned', it will make the the acl still belong to that anonymous owner in  'RGWAccessControlPolicy_S3::create_canned'.  I think We should not only belong the `grant_map` to the bucket owner but also should belong the `acl` to the owner.
So should we check the anonymous in the  `RGWAccessControlPolicy_S3::create_canned`? We can make a local copy with `ACLOwner bucket_owner` for  the input of `RGWAccessControlPolicy_S3::create_canned` and deeper pass it in `RGWAccessControlList_S3::create_canned` or we can just pass the `bucket_owner` as the input value of `ACLOwner& owner` of `RGWAccessControlPolicy_S3::create_canned`. Which one do you think is better?  Thanks.